### PR TITLE
[inductor] Persistent reductions

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #92267

This one may need to wait for the new MLIR Triton to land as it triggers some Triton crashes.

Before:
```
$ pytest test/inductor/test_torchinductor.py -vsk test_softmax_one_kernel_loop_cuda
...
@reduction(
    size_hints=[16, 32],
    reduction_hint=ReductionHint.INNER,
    filename=__file__,
    meta={'signature': {0: '*fp32', 1: '*fp32', 2: 'i32', 3: 'i32'}, 'device': 0, 'constants': {}, 'mutated_arg_names': [], 'configs': [instance_descriptor(divisible_by_16=(0, 1, 2, 3), equal_to_1=())]}
)
@triton.jit
def triton_(in_ptr0, out_ptr2, xnumel, rnumel, XBLOCK : tl.constexpr, RBLOCK : tl.constexpr):
    xnumel = 16
    rnumel = 32
    xoffset = tl.program_id(0) * XBLOCK
    xindex = xoffset + tl.arange(0, XBLOCK)[:, None]
    xmask = xindex < xnumel
    rbase = tl.arange(0, RBLOCK)[None, :]
    x0 = xindex
    _tmp1 = tl.zeros([XBLOCK, RBLOCK], tl.float32) + float("-inf")
    for roffset in range(0, rnumel, RBLOCK):
        rindex = roffset + rbase
        rmask = rindex < rnumel
        r1 = rindex
        tmp0 = tl.load(in_ptr0 + (r1 + (32*x0)), rmask & xmask, eviction_policy='evict_last')
        _tmp1 = tl.where(xmask & rmask & (_tmp1 < tmp0), tmp0, _tmp1)
    tmp1 = tl.max(_tmp1, 1)[:, None]
    _tmp5 = tl.zeros([XBLOCK, RBLOCK], tl.float32) + 0
    for roffset in range(0, rnumel, RBLOCK):
        rindex = roffset + rbase
        rmask = rindex < rnumel
        r1 = rindex
        tmp2 = tl.load(in_ptr0 + (r1 + (32*x0)), rmask & xmask, eviction_policy='evict_last')
        tmp3 = tmp2 - tmp1
        tmp4 = tl.exp(tmp3)
        _tmp5 = tl.where(xmask & rmask, _tmp5 + tmp4, _tmp5)
    tmp5 = tl.sum(_tmp5, 1)[:, None]
    for roffset in range(0, rnumel, RBLOCK):
        rindex = roffset + rbase
        rmask = rindex < rnumel
        r1 = rindex
        tmp6 = tl.load(in_ptr0 + (r1 + (32*x0)), rmask & xmask, eviction_policy='evict_last')
        tmp7 = tmp6 - tmp1
        tmp8 = tl.exp(tmp7)
        tmp9 = tmp8 / tmp5
        tl.store(out_ptr2 + (r1 + (32*x0) + tl.zeros([XBLOCK, RBLOCK], tl.int32)), tmp9, rmask & xmask)
```

After
```
$ pytest test/inductor/test_torchinductor.py -vsk test_softmax_one_kernel_persist_cuda
...
@persistent_reduction(
    size_hints=[16, 32],
    reduction_hint=ReductionHint.INNER,
    filename=__file__,
    meta={'signature': {0: '*fp32', 1: '*fp32', 2: 'i32', 3: 'i32'}, 'device': 0, 'constants': {}, 'mutated_arg_names': [], 'configs': [instance_descriptor(divisible_by_16=(0, 1, 2, 3), equal_to_1=())]}
)
@triton.jit
def triton_(in_ptr0, out_ptr2, xnumel, rnumel, XBLOCK : tl.constexpr, RBLOCK : tl.constexpr):
    xnumel = 16
    rnumel = 32
    xoffset = tl.program_id(0) * XBLOCK
    xindex = xoffset + tl.arange(0, XBLOCK)[:, None]
    xmask = xindex < xnumel
    rindex = tl.arange(0, RBLOCK)[None, :]
    rmask = rindex < rnumel
    r1 = rindex
    x0 = xindex
    tmp0 = tl.load(in_ptr0 + (r1 + (32*x0)), rmask & xmask)
    tmp2 = tl.where(xmask & rmask, tmp0, float("-inf"))
    tmp3 = tl.max(tmp2, 1)[:, None]
    tmp4 = tmp0 - tmp3
    tmp5 = tl.exp(tmp4)
    tmp7 = tl.where(xmask & rmask, tmp5, 0)
    tmp8 = tl.sum(tmp7, 1)[:, None]
    tmp9 = tmp5 / tmp8
    tl.store(out_ptr2 + (r1 + (32*x0) + tl.zeros([XBLOCK, RBLOCK], tl.int32)), tmp9, rmask & xmask)
```

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: test/inductor/test_torchinductor.py [hunk 1]
```diff
@@ -2602,6 +2603,18 @@ def fn(x):
         if self.device != "cpu":
             self.assertEqual(torch._inductor.metrics.generated_kernel_count, 1)
 
+    @patch.object(config.triton, "persistent_reductions", False)
+    def test_softmax_one_kernel_loop(self):
+        def fn(x):
+            x_max = torch.amax(x, 1, keepdim=True)
+            unnormalized = torch.exp(x * x_max)
```
Comment (vadimkantorov) [inline test/inductor/test_torchinductor.py:2610]: should here be `x - x_max`?

Comment (jansel) [reply test/inductor/test_torchinductor.py:2610]: Yes, thanks.

**Comment by Chillee** [comment]
Do you have any benchmarks for this?

**Comment by jansel** [comment]
> Do you have any benchmarks for this?

Yeah, I have some RTX 3090 numbers that show some decent speedups (lots of ~1-5% wins spread out across most models).  Need to rerun on an A100 though.

**Comment by jansel** [comment]
There is some performance data with this [enabled](https://github.com/pytorch/pytorch/actions/runs/4068616915/jobs/7007782583) and [disabled](https://github.com/pytorch/pytorch/actions/runs/4071185125/jobs/7012954316).

It is kind of hard to parse the results from the logs and it doesn't seem like the `.csv` output is captured, but the summaries are:

Torchbench
```
Enabled:  gmean=1.36x mean=1.41x
Disabled: gmean=1.35x mean=1.38x
```

Huggingface:
```
Enabled:  gmean=1.61x mean=1.64x
Disabled: gmean=1.59x mean=1.62x
```


Timm part 1
```
Enabled:  gmean=1.39x mean=1.41x
Disabled: gmean=1.39x mean=1.40x
```

Timm part 2
```
Enabled:  gmean=1.35x mean=1.38x
Disabled: gmean=1.34x mean=1.35x
```

**Comment by Chillee** [comment]
@jansel Would be cool to see some performance results on microbenchmarks like layer norm or softmax :)

**Comment by lezcano** [comment]
You can get the `.csv` results in the `test-reports*` artifacts in https://github.com/pytorch/pytorch/actions/runs/4068616915.

File: test/inductor/test_torchinductor.py [hunk 2]
```diff
@@ -2771,18 +2771,31 @@ def fn(a, b):
         if self.device != "cpu":
             self.assertEqual(torch._inductor.metrics.generated_kernel_count, 1)
 
-    def test_softmax_one_kernel(self):
+    @patch.object(config.triton, "persistent_reductions", True)
```
Comment (Chillee) [inline test/inductor/test_torchinductor.py:2774]: Might be worth adding one of the "triton-code based" tests that @eellison added some infra for: https://github.com/pytorch/pytorch/pull/93810/files#diff-809c39aeafb3acc92289f42a63e670a8719d4ce5627d5f88820142d80edf8d2aR6610

File: torch/_inductor/codegen/triton.py [hunk 3]
```diff
@@ -574,6 +584,28 @@ def simplify_indexing(index: sympy.Expr):
...
+        if not (self.inside_reduction and config.triton.persistent_reductions):
+            return False
+        threshold = {
+            ReductionHint.INNER: 1024,
+        }.get(self.reduction_hint, 64)
+        hint = V.graph.sizevars.size_hint(self.numels[-1])
+        if hint > threshold:
+            return False
+
+        # if self.reduction_hint == ReductionHint.INNER and hint < 128:
```
Comment (Chillee) [inline torch/_inductor/codegen/triton.py:601]: Clean up?

File: torch/_inductor/config.py [hunk 4]
```diff
@@ -175,6 +175,9 @@ class triton:
     # should we put op names in kernel names
     descriptive_kernel_names = False
 
+    # use alternate codegen for smaller reductions
+    persistent_reductions = False
```
Comment (Chillee) [inline torch/_inductor/config.py:179]: Should this be on by default?

**Review by Chillee — APPROVED** [review]
Makes sense to me!

**Review by Chillee — COMMENTED** [review]
Also, I was curious how this affected my softmax microbenchmark. Unsurprisingly, it has no effect, since persistent reductions only apply when the reduction range is < 1024, where we're already generating bandwidth-bound kernels.

But.... if we increase the heuristic to 
```
threshold = {
    ReductionHint.INNER: 16384*4, # (up from 1024)
}.get(self.reduction_hint, 64)
```
We get this plot.
image

eager = eager perf for `torch.softmax`
Inductor = existing performance with persistent reductions off
heuristic = autotuning non-persistent reductions across many block sizes.
persistent = the above heuristic with persistent reductions.

TL;DR: Persistent reductions seem to allow us to much more reliably fit things into shared memory, as opposed to non-persistent reductions, where we required autotuning in order to stay close to peak bandwidth in the 1e3 - 1e4 range. 

So... future todo might be to more aggressively use persistent reductions depending on whether we can fit them (cc: @lezcano ).

**Comment by Chillee** [comment]
Another thing that would be nice is being able to autotune between persistent and non-persistent reductions, which seems kind of difficult to add in the current structure. I guess maybe it’s possible to generate both and then autotune between those.

**Comment by lezcano** [comment]
Related to the heuristics above, in https://github.com/pytorch/pytorch/pull/90994 I implemented a poorman's (upper bound) for the the number of registers necessary to implement a kernel to guess the maximum loop unrolling we can do in a kernel. With that, for softmax, I was getting even better numbers than the minimum of the graphs shown by Horace above (see https://github.com/pytorch/pytorch/pull/90994#issuecomment-1354587320).
Now, when I run the benchmarks, I'm getting a regression, so this, apparently, not good enough. If you have any ideas what could be going on there, please comment. Is there any other way to get a better a priori estimate for the number of registers a kernel may use?
My next steps would be to get one of the models it regresses, find a kernel it regresses and run it through nsight to see the heck is going on with the register usage in that kernel.

Also, now that we are into this, what I figured that could be done in the future is a scheme as follows:
- You estimate the max parallelism you can get. 
- You compile a kernel with those params. 
- When this kernel is executed, you have a look at the registers it used, occupancy and a few other parameters and you guard on things like "not spilling registers" and "not having too low occupancy provided you don't spill registers"
- If you think you can compile a better kernel in the next run, you do so and look at those parameters again.

This would make compilation slower, but the final result would be better, so this could be done in `max-autotune`. If this sounds interesting, I'd be keen on having a go at it, but I'd need to discuss the where-goes-what with one of you guys.

File: torch/_inductor/codegen/triton.py [hunk 5]
```diff
@@ -574,6 +584,28 @@ def simplify_indexing(index: sympy.Expr):
...
 
+    def should_use_persistent_reduction(self):
+        """
+        Heuristic to set self.persistent_reduction and add guards
+        if needed.
+        """
+        if not (self.inside_reduction and config.triton.persistent_reductions):
+            return False
+        threshold = {
+            ReductionHint.INNER: 1024,
```
[the remark this answers is not in the document]
Comment (jansel) [reply torch/_inductor/codegen/triton.py:599]: I like this idea

File: torch/_inductor/config.py (hunk 4 above)
Comment (jansel) [reply torch/_inductor/config.py:182]: Going to leave this off as this is triggering some rare triton errors.

**Comment by bozhiyou** [comment]
Could anyone explain why the `After` code is considered *persistent* rather than `Before`? It confuses me that `_tmp` in `Before` seems to be the persistent part.

> Before:
> 
> ```
> $ pytest test/inductor/test_torchinductor.py -vsk test_softmax_one_kernel_loop_cuda
> ...
> @reduction(
>     size_hints=[16, 32],
>     reduction_hint=ReductionHint.INNER,
>     filename=__file__,
>     meta={'signature': {0: '*fp32', 1: '*fp32', 2: 'i32', 3: 'i32'}, 'device': 0, 'constants': {}, 'mutated_arg_names': [], 'configs': [instance_descriptor(divisible_by_16=(0, 1, 2, 3), equal_to_1=())]}
> )
> @triton.jit
> def triton_(in_ptr0, out_ptr2, xnumel, rnumel, XBLOCK : tl.constexpr, RBLOCK : tl.constexpr):
>     xnumel = 16
>     rnumel = 32
>     xoffset = tl.program_id(0) * XBLOCK
>     xindex = xoffset + tl.arange(0, XBLOCK)[:, None]
>     xmask = xindex < xnumel
>     rbase = tl.arange(0, RBLOCK)[None, :]
>     x0 = xindex
>     _tmp1 = tl.zeros([XBLOCK, RBLOCK], tl.float32) + float("-inf")
>     for roffset in range(0, rnumel, RBLOCK):
>         rindex = roffset + rbase
>         rmask = rindex < rnumel
>         r1 = rindex
>         tmp0 = tl.load(in_ptr0 + (r1 + (32*x0)), rmask & xmask, eviction_policy='evict_last')
>         _tmp1 = tl.where(xmask & rmask & (_tmp1 < tmp0), tmp0, _tmp1)
>     tmp1 = tl.max(_tmp1, 1)[:, None]
>     _tmp5 = tl.zeros([XBLOCK, RBLOCK], tl.float32) + 0
>     for roffset in range(0, rnumel, RBLOCK):
>         rindex = roffset + rbase
>         rmask = rindex < rnumel
>         r1 = rindex
>         tmp2 = tl.load(in_ptr0 + (r1 + (32*x0)), rmask & xmask, eviction_policy='evict_last')
>         tmp3 = tmp2 - tmp1
>         tmp4 = tl.exp(tmp3)
>         _tmp5 = tl.where(xmask & rmask, _tmp5 + tmp4, _tmp5)
>     tmp5 = tl.sum(_tmp5, 1)[:, None]
>     for roffset in range(0, rnumel, RBLOCK):
>         rindex = roffset + rbase
>         rmask = rindex < rnumel
>         r1 = rindex
>         tmp6 = tl.load(in_ptr0 + (r1 + (32*x0)), rmask & xmask, eviction_policy='evict_last')
>         tmp7 = tmp6 - tmp1
>         tmp8 = tl.exp(tmp7)
>         tmp9 = tmp8 / tmp5
>         tl.store(out_ptr2 + (r1 + (32*x0) + tl.zeros([XBLOCK, RBLOCK], tl.int32)), tmp9, rmask & xmask)
> ```
> 
> After
> 
> ```
> $ pytest test/inductor/test_torchinductor.py -vsk test_softmax_one_kernel_persist_cuda
> ...
> @persistent_reduction(
>     size_hints=[16, 32],
>     reduction_hint=ReductionHint.INNER,
>     filename=__file__,
>     meta={'signature': {0: '*fp32', 1: '*fp32', 2: 'i32', 3: 'i32'}, 'device': 0, 'constants': {}, 'mutated_arg_names': [], 'configs': [instance_descriptor(divisible_by_16=(0, 1, 2, 3), equal_to_1=())]}
> )
> @triton.jit
> def triton_(in_ptr0, out_ptr2, xnumel, rnumel, XBLOCK : tl.constexpr, RBLOCK : tl.constexpr):
>     xnumel = 16
>     rnumel = 32
>     xoffset = tl.program_id(0) * XBLOCK
>     xindex = xoffset + tl.arange(0, XBLOCK)[:, None]
>     xmask = xindex < xnumel
>     rindex = tl.arange(0, RBLOCK)[None, :]
>     rmask = rindex < rnumel
>     r1 = rindex
>     x0 = xindex
>     tmp0 = tl.load(in_ptr0 + (r1 + (32*x0)), rmask & xmask)
>     tmp2 = tl.where(xmask & rmask, tmp0, float("-inf"))
>     tmp3 = tl.max(tmp2, 1)[:, None]
>     tmp4 = tmp0 - tmp3
>     tmp5 = tl.exp(tmp4)
>     tmp7 = tl.where(xmask & rmask, tmp5, 0)
>     tmp8 = tl.sum(tmp7, 1)[:, None]
>     tmp9 = tmp5 / tmp8
>     tl.store(out_ptr2 + (r1 + (32*x0) + tl.zeros([XBLOCK, RBLOCK], tl.int32)), tmp9, rmask & xmask)
> ```
Any clarification is appreciated in advance :)
